ggml/src/ggml-vulkan/ggml-vulkan.cpp

```diff
@@ -1418,11 +1419,11 @@ static void ggml_vk_load_shaders(vk_device& device) {
 
         l_warptile = { 128, 128, 128, 16, device->subgroup_size * 2, 64, 2, tm_l, tn_l, tk_l, device->subgroup_size };
         m_warptile = { 128,  64,  64, 16, device->subgroup_size, 32, 2, tm_m, tn_m, tk_m, device->subgroup_size };
-        s_warptile = { subgroup_size_16, 32, 32, 16, 32, 32, 2, tm_s, tn_s, tk_s, device->subgroup_size };
+        s_warptile = { subgroup_size_32, 32, 32, 16, 32, 32, 2, tm_s, tn_s, tk_s, device->subgroup_size };
```

Comment:
What goes wrong when this is smaller than 32? Does it depend on tm/tn/tk?

Does llavapipe actually support KHR_cooperative_matrix? What tile sizes does it advertise?

Reply:
If this is smaller than 32 I get a crash on AMD and incorrect results on llvmpipe.

I originally thought that this was due to BM but I wasn't able to figure it out from looking at the mul mat algorithm. As this looked like a subgroup size issue I experimented with `test_backend_ops` and inference runs using different subgroup sizes. With the block size set to 16 it didn't work properly regardless of what I set the warp size to, but if set to 32 it worked properly with warp sizes <= 32. The desktop GPUs this Vulkan backend was written on have subgroup sizes of at least 32 so something like this would be easily missed unless you specifically override `device->subgroup_size` during testing.

Llvmpipe doesn't support KHR_cooperative_matrix and the coopmat paths aren't tested here.

Reply:
Is AMD advertising a subgroup size smaller than 32?

Reply:
Nope the hardware is still 64. I'm just overriding the warptile warp size or the entire `device->subgroup_size` to test how the backend reacts to a smaller subgroup size. We don't use subgroup specific operations so this shouldn't be an issue and I am able to run inference successfully on AMD with `device->subgroup_size` set to 16 using my PR.

Reply:
Oops this was a mistake and setting the subgroup size to 32 is only necessary for `s_warptile_mmq` and not for `s_warptile`. After looking into the algorithm in more detail I think this is due to BK being set to 32 when working with a quantized A.

https://github.com/netrunnereve/llama.cpp/blob/663716ce4229fe9880b4190809199730550acd16/ggml/src/ggml-vulkan/vulkan-shaders/mul_mm.comp#L69C1-L69C90

Reply:
I suspect it's due to this code, which would end up computing a stride of zero if the workgroupsize is smaller than BK:
```
    const uint loadstride_a = gl_WorkGroupSize.x * LOAD_VEC_A / BK;
    const uint loadstride_b = gl_WorkGroupSize.x * LOAD_VEC_B / BK;
```

So I agree with rounding up to 32.

Reply:
That's what I initially thought as well but LOAD_VEC_A is 2 for quant types and LOAD_VEC B is either 4 or 8. As a result a workgroup size of 16 and a BK of 32 shouldn't cause a 0 stride.

Reply:
LOAD_VEC_B is only 4/8 for the vector-aligned shader. The version with boundary checks uses LOAD_VEC_B=1.